File: src/libGEjava/Mapper.java
/*
 * Mapper.java
 *
 * Created on June 1, 2007, 11:24 PM
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

/**
 *
 * @author adilraja
 */
package libGEjava;

import java.util.*;
import java.util.ArrayList;

public abstract class Mapper extends ArrayList<Rule> {
    
    protected Genotype genotype;
    protected Phenotype phenotype;
    protected abstract boolean genotype2phenotype();
    protected abstract boolean phenotype2genotype();
    
    public Mapper(){
        super();
        this.phenotype=new Phenotype();
        this.genotype=new Genotype();
        
    }
    /**
     *Constructor with genotype structure. set to new newGenotype
     */
    public Mapper(final Genotype newGenotype){
        super();
        this.genotype=newGenotype;
    }
    /**
     *Constructor with phenotype structor. set phenotype to newPhenotype
     */
    public Mapper(final Phenotype newPhenotype){
        super();
        this.phenotype=new Phenotype(newPhenotype);
    }
    /**
     *Copy Constructor
     */
     public Mapper(final Mapper copy){
         super(copy);
         this.genotype=new Genotype(copy.genotype);
         this.phenotype=new Phenotype(copy.phenotype);
     }
     /**
      *returns the genotype structure of this mapper
      */
     public final Genotype getGenotype(){
         return this.genotype;
     }
     /**
      *Sets this mapper's genotype structure to be a copy of the argument, newgenotype,
      *and calls the genotype2phenotype private method.
      */
    public void setGenotype(final Genotype newGenotype){
        this.genotype=new Genotype(newGenotype);
        genotype2phenotype();
    }
    /**
     *returns a pointer to the phenotype structure of this mapper
     */
    public final Phenotype getPhenotype(){
        return this.phenotype;
    }
    /**
     *Sets this mapper's phenotype structure to be a copy of the argument phenotype
     *structure, and calls the phenotype2genotype private method
     */
    public void setPhenotype(final Phenotype newPhenotype){
        phenotype=new Phenotype(newPhenotype);
        phenotype2genotype();
    }
    /**
     *Sets the maximum codon value of the genotype structure
     */
    public void setGenotypeMaxCodonValue(final int newMaxCodonValue){//CodonType=int
        this.genotype.setMaxCodonValue(newMaxCodonValue);
    }
    //something left here
}
// file complete

File: libGEC++/libGE-0.26/src/Mapper.cpp
// Mapper.cpp -*- C++ -*-
#ifndef _MAPPER_CPP_
#define _MAPPER_CPP_

//#include <config.h>

#include<iostream>

#include "Mapper.h"

///////////////////////////////////////////////////////////////////////////////
// Default constructor.
Mapper::Mapper(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::Mapper()' called\n";
#endif
}

///////////////////////////////////////////////////////////////////////////////
// Constructor with Genotype structure; set Genotype to newGenotype.
Mapper::Mapper(const Genotype &newGenotype):genotype(newGenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::Mapper(const Genotype &)' called\n";
#endif
}

///////////////////////////////////////////////////////////////////////////////
// Constructor with Phenotype structure; set Phenotype to newPhenotype.
Mapper::Mapper(const Phenotype &newPhenotype):phenotype(newPhenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::Mapper(const Phenotype &)' called\n";
#endif
}

///////////////////////////////////////////////////////////////////////////////
// Copy constructor.
Mapper::Mapper(const Mapper& copy):genotype(copy.genotype),phenotype(copy.phenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::Mapper(const Mapper&)' called\n";
#endif
}

///////////////////////////////////////////////////////////////////////////////
// Destructor.
Mapper::~Mapper(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::~Mapper()' called\n";
#endif
}

///////////////////////////////////////////////////////////////////////////////
// Returns a pointer to the genotype structure of this mapper.
Genotype const * Mapper::getGenotype() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::getGenotype() const' called\n";
#endif
	return &genotype;
}

///////////////////////////////////////////////////////////////////////////////
// Sets this mapper's genotype structure to be a copy of the argument genotype
// structure, and calls the genotype2Phenotype private method.
void Mapper::setGenotype(const Genotype &newGenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::setGenotype(const Genotype &)' called\n";
#endif
	genotype=newGenotype;
	genotype2Phenotype();
}

///////////////////////////////////////////////////////////////////////////////
// Returns apointer to the phenotype structure of this mapper.
Phenotype const * Mapper::getPhenotype() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::getPhenotype() const' called\n";
#endif
	return &phenotype;
}

///////////////////////////////////////////////////////////////////////////////
// Sets this mapper's phenotype structure to be a copy of the argument
// phenotype structure, and calls the phenotype2Genotype private method.
void Mapper::setPhenotype(const Phenotype &newPhenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::setPhenotype(const Phenotype &)' called\n";
#endif
	phenotype=newPhenotype;
	phenotype2Genotype();
}

///////////////////////////////////////////////////////////////////////////////
// Sets the maximum codon value of the genotype structure.
void Mapper::setGenotypeMaxCodonValue(const CodonType newMaxCodonValue){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Mapper::setGenotypeMaxCodonValue(const CodonType)' called\n";
#endif
	genotype.setMaxCodonValue(newMaxCodonValue);
}

#endif

